File: 29Items-Projects/7Algorithms/5-DIVIDE-CONQUER/src/main/python/divide_conquer/algorithms/maximum_subarray.py
"""Divide-and-conquer maximum subarray."""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class Result:
    start: int
    end: int
    total: int


def find(values: list[int]) -> Result:
    """Return the maximum-sum contiguous subarray."""
    if not values:
        raise ValueError("values must not be empty")
    return _find(values, 0, len(values) - 1)


def _find(values: list[int], left: int, right: int) -> Result:
    if left == right:
        return Result(left, right, values[left])

    mid = left + (right - left) // 2
    left_best = _find(values, left, mid)
    right_best = _find(values, mid + 1, right)
    crossing_best = _crossing(values, left, mid, right)
    return max(left_best, right_best, crossing_best, key=lambda result: result.total)


def _crossing(values: list[int], left: int, mid: int, right: int) -> Result:
    best_left_sum = -10**18
    current = 0
    best_left = mid
    for index in range(mid, left - 1, -1):
        current += values[index]
        if current > best_left_sum:
            best_left_sum = current
            best_left = index

    best_right_sum = -10**18
    current = 0
    best_right = mid + 1
    for index in range(mid + 1, right + 1):
        current += values[index]
        if current > best_right_sum:
            best_right_sum = current
            best_right = index

    return Result(best_left, best_right, best_left_sum + best_right_sum)

File: 29Items-Projects/7Algorithms/5-DIVIDE-CONQUER/src/main/java/com/divideconquer/algorithms/MaximumSubarray.java
package com.divideconquer.algorithms;

public final class MaximumSubarray {
    private MaximumSubarray() {
    }

    public static final class Result {
        private final int startInclusive;
        private final int endInclusive;
        private final int sum;

        public Result(int startInclusive, int endInclusive, int sum) {
            this.startInclusive = startInclusive;
            this.endInclusive = endInclusive;
            this.sum = sum;
        }

        public int startInclusive() {
            return startInclusive;
        }

        public int endInclusive() {
            return endInclusive;
        }

        public int sum() {
            return sum;
        }

        @Override
        public String toString() {
            return "Result{start=" + startInclusive + ", end=" + endInclusive + ", sum=" + sum + "}";
        }
    }

    public static Result find(int[] values) {
        if (values == null || values.length == 0) {
            throw new IllegalArgumentException("values must not be empty");
        }
        return find(values, 0, values.length - 1);
    }

    private static Result find(int[] values, int left, int right) {
        if (left == right) {
            return new Result(left, right, values[left]);
        }

        int mid = left + (right - left) / 2;
        Result leftBest = find(values, left, mid);
        Result rightBest = find(values, mid + 1, right);
        Result crossingBest = crossing(values, left, mid, right);
        return maxBySum(leftBest, rightBest, crossingBest);
    }

    private static Result crossing(int[] values, int left, int mid, int right) {
        int bestLeftSum = Integer.MIN_VALUE;
        int sum = 0;
        int bestLeft = mid;
        for (int index = mid; index >= left; index--) {
            sum += values[index];
            if (sum > bestLeftSum) {
                bestLeftSum = sum;
                bestLeft = index;
            }
        }

        int bestRightSum = Integer.MIN_VALUE;
        sum = 0;
        int bestRight = mid + 1;
        for (int index = mid + 1; index <= right; index++) {
            sum += values[index];
            if (sum > bestRightSum) {
                bestRightSum = sum;
                bestRight = index;
            }
        }

        return new Result(bestLeft, bestRight, bestLeftSum + bestRightSum);
    }

    private static Result maxBySum(Result first, Result second, Result third) {
        Result best = first.sum() >= second.sum() ? first : second;
        return best.sum() >= third.sum() ? best : third;
    }
}
